### PR TITLE
wrapQtAppsHook: wrap binaries in libexec

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
See feedback at https://github.com/NixOS/nixpkgs/pull/54525#issuecomment-511164934.
This should be fine to do, it's already done in `wrapGAppsHook`.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by worldofpeace** [comment]
I've noticed with `wrapGAppsHook` that if there's binaries at `sbin` that they don't get wrapped.
I'm thinking this was because `move-sbin` `fixupOutputHooks` came after `wrapGAppsHook` so they don't get touched.

Mentioning because I'd think `wrapQtAppsHook` could inherit that issue.

**Comment by ttuegel** [comment]
What happens when a `bin/` executable execs a `libexec/` executable? Won't the environment grow?

**Comment by worldofpeace** [comment]
> What happens when a `bin/` executable execs a `libexec/` executable? Won't the environment grow?

Yes, environment variables are going to propagate to child processes and often they can be wrapped too.
So this leakage can be problematic, but it's a known drawback to this approach.

**Review by ttuegel — APPROVED** [review]

**Comment by coreyoconnor** [comment]
LGTM. This resolved the activitymanagerd failure I noted in https://github.com/NixOS/nixpkgs/pull/54525#issuecomment-511164934

However, the plasma 5 test is still failing. A different failure. Does not appear to be a regression caused by this change. Log:

* https://gist.github.com/coreyoconnor/410e5291aebce30c611f68abcf448870

Not clear to me what failed this time.

**Comment by worldofpeace** [comment]
> LGTM. This resolved the activitymanagerd failure I noted in [#54525 (comment)](https://github.com/NixOS/nixpkgs/pull/54525#issuecomment-511164934)
> 
> However, the plasma 5 test is still failing. A different failure. Does not appear to be a regression caused by this change. Log:
> 
>     * https://gist.github.com/coreyoconnor/410e5291aebce30c611f68abcf448870
> 
> 
> Not clear to me what failed this time.

Can you run the test interactively and debug from there?
```
nix-build nixos/tests/plasma5.nix -A driver 
```

**Comment by worldofpeace** [comment]
Running the test interactively everything works for me
Screenshot from 2019-07-16 18 44 16


What the issue here is that the wrappers fuzzed the `--argv0` so the test needs to be adapted
* https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/plasma5.nix#L53

I'll include a commit for that here.

**Comment by worldofpeace** [comment]
Ok, after inspecting `dolphin` it's not being wrapped so it gets that error at runtime...
Pretty confused by that.

**Comment by worldofpeace** [comment]
I've discovered that `isQtApp` wasn't returning a non-zero exit status for dolphin's binaries.
Decided to just remove it https://github.com/NixOS/nixpkgs/pull/64720/commits/4908d382ae1eb9f618ea8745cb6794db1dba9520

**Comment by worldofpeace** [comment]
So the NixOS test now "passes", but I'm seeing errors in the output like

```
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-alice'
```

among others that somehow get drawn to the desktop after I run it interactively
Screenshot from 2019-07-16 23 03 13

**Comment by coreyoconnor** [comment]
Thanks! I can confirm the test passes. I haven't tried running the driver interactively. If I get a chance I'll try it and see.

**Comment by FRidh** [comment]
@worldofpeace please remove the WIP label when it's not applicable.

**Comment by worldofpeace** [comment]
@ttuegel Was `isQtApp` meant to prevent scripts from being wrapped?

I'm thinking that the new issues I'm seeing are because of that happening.
Though it's still possible that a script could need to be wrapped with $XDG* variables.

**Comment by ttuegel** [comment]
@worldofpeace `isQtApp` was intended to prevent wrapping scripts and non-GUI binaries that may not even depend on Qt. If it's causing problems, then remove it.

File: pkgs/development/libraries/qt-5/hooks/wrap-qt-apps-hook.sh [hunk 1]
```diff
@@ -86,8 +82,6 @@ wrapQtAppsHook() {
 
         find "$targetDir" -executable -print0 | while IFS= read -r -d '' file
         do
-            isQtApp "$file" || continue
```
Comment (Mic92) [inline pkgs/development/libraries/qt-5/hooks/wrap-qt-apps-hook.sh]: `isELF` could replace this.

Comment (worldofpeace) [reply pkgs/development/libraries/qt-5/hooks/wrap-qt-apps-hook.sh]: Was just about to submit that

**Comment by worldofpeace** [comment]
So with the latest commit, plasma5 test succeeds for me locally and interactively nothing seems wrong with the session.

I'm going to merge this to master so the channels can get unblocked.

**Comment by FRidh** [comment]
https://github.com/NixOS/nixpkgs/pull/64720/commits/7d6ab0ad7a5a5b1765a6d70b5df861d14e3c3482 breaks Python scripts that use Qt5 because they don't get wrapped because of the `isELF` check.

**Comment by coreyoconnor** [comment]
I have patches in my nixpkgs that may impact this. Adding the issue I'm encountering here just in case it's actually related to this change. Plus, takes a while to build a variation. ;)

Some QT apps, for instance QGIS and OpenSCAD, encounter an infinite recursion that has a stack trace like so:

~~~
#0  0x000015554e001680 in QWindow::flags() const () from /nix/store/v4s0nq286hxwjms4k1sd378n597l3rka-qtbase-5.12.3/lib/libQt5Gui.so.5
#1  0x000015554e0016b9 in QWindow::type() const () from /nix/store/v4s0nq286hxwjms4k1sd378n597l3rka-qtbase-5.12.3/lib/libQt5Gui.so.5
#2  0x000015553e630c61 in ?? () from /nix/store/35855h0lfnpr21xbs48kd1ianwzzw0yv-qtbase-5.12.0-bin/lib/qt-5.12/plugins/xcbglintegrations/libqxcb-glx-integration.so
#3  0x000015554e001682 in QWindow::flags() const () from /nix/store/v4s0nq286hxwjms4k1sd378n597l3rka-qtbase-5.12.3/lib/libQt5Gui.so.5
#4  0x000015554e0016b9 in QWindow::type() const () from /nix/store/v4s0nq286hxwjms4k1sd378n597l3rka-qtbase-5.12.3/lib/libQt5Gui.so.5
#5  0x000015553e630c61 in ?? () from /nix/store/35855h0lfnpr21xbs48kd1ianwzzw0yv-qtbase-5.12.0-bin/lib/qt-5.12/plugins/xcbglintegrations/libqxcb-glx-integration.so
#6  0x000015554e001682 in QWindow::flags() const () from /nix/store/v4s0nq286hxwjms4k1sd378n597l3rka-qtbase-5.12.3/lib/libQt5Gui.so.5
#7  0x000015554e0016b9 in QWindow::type() const () from /nix/store/v4s0nq286hxwjms4k1sd378n597l3rka-qtbase-5.12.3/lib/libQt5Gui.so.5
#8  0x000015553e630c61 in ?? () from /nix/store/35855h0lfnpr21xbs48kd1ianwzzw0yv-qtbase-5.12.0-bin/lib/qt-5.12/plugins/xcbglintegrations/libqxcb-glx-integration.so
~~~

Note the `libqxcb-glx-integration` library.

(This is the issue I mentioned running qt apps in #63829 .)

**Comment by coreyoconnor** [comment]
> I have patches in my nixpkgs that may impact this. Adding the issue I'm encountering here just in case it's actually related to this change. Plus, takes a while to build a variation. ;)

I've reverted to `nixos-unstable` plus a few necessary patches and am not able to reproduce the above issue. I'll be adding back in QT PRs to see if they repro the issue.

**Pull request merged**